ament_export_interfaces assumes export name and target name are the same

Blocks #178

I would expect `target_link_libraries(my_target ${a_dependency_INTERFACES})` to link `my_target` against all targets exported by `a_dependency` if it used `ament_export_interfaces()`, but currently this only works if the "export" name is the same as the target name. The variable `..._INTERFACES` is unusable if that assumption is violated.

This likely to happen because the `ament_export_interfaces` also [expects users to call `install(TARGETS ... EXPORT name )`](https://github.com/ament/ament_cmake/blob/e1ff1c1a0a1e08d43e939cdb943a88be601808bd/ament_cmake_export_interfaces/cmake/ament_export_interfaces.cmake#L18-L19) which is what gives the "export" it's name. For example, it looks like [`rosbag2_transport` violates this assumption](https://github.com/ros2/rosbag2/blob/4bb9d892154c15fc2670891178ede003f82c6a46/rosbag2_transport/CMakeLists.txt#L74).

1. `ament_export_interfaces()` [expects a list of "export" names as arguments.](https://github.com/ament/ament_cmake/blob/e1ff1c1a0a1e08d43e939cdb943a88be601808bd/ament_cmake_export_interfaces/cmake/ament_export_interfaces.cmake#L25)
1. The `ament_package()` hook uses those names to call [`install(EXPORT ...)` to install the "export"s](https://github.com/ament/ament_cmake/blob/e1ff1c1a0a1e08d43e939cdb943a88be601808bd/ament_cmake_export_interfaces/cmake/ament_cmake_export_interfaces_package_hook.cmake#L28-L32). This is fine.
1. the `<project>Config.cmake` hook uses those names to [`include()` the installed exports](https://github.com/ament/ament_cmake/blob/e1ff1c1a0a1e08d43e939cdb943a88be601808bd/ament_cmake_export_interfaces/cmake/ament_cmake_export_interfaces-extras.cmake.in#L8). This is also fine.
1. the `<project>Config.cmake` hook creates a variable `${PROJECT_NAME}_INTERFACES` containing `<project>::<export>`. This is a problem
    * It's using a double colon, [which suggests it's the target name](https://cmake.org/cmake/help/v3.0/policy/CMP0028.html), but the "export" is not a target and can't be linked against.
    * Downstream packages linking against `${..._INTERFACES}` will get an error like:
        ```
        CMake Error at CMakeLists.txt:11 (add_executable):
          Target "my_target" links to target "a_dependency::export_some_target" but the target was not
          found.  Perhaps a find_package() call is missing for an IMPORTED target, or
          an ALIAS target is missing?
        ```

It seems likely that any fix for this will involve changing the `ament_export_interfaces()` api. One option is to give it the targets that have been installed in an "export" so it can append those to the `..._INTERFACES` variable.

--- a/ament_cmake_export_interfaces/cmake/ament_cmake_export_interfaces-extras.cmake.in
+++ b/ament_cmake_export_interfaces/cmake/ament_cmake_export_interfaces-extras.cmake.in
@@ -9,17 +9,16 @@ if(NOT _exported_interfaces STREQUAL "")
     include("${_export_file}")
 
     # extract the target names associated with the export
+    set(_regex "foreach\\(_expectedTarget (.+)\\)")
     file(
       STRINGS "${_export_file}" _foreach_targets
-      REGEX "foreach\\(_expectedTarget (.+)\\)")
+      REGEX "${_regex}")
     list(LENGTH _foreach_targets _matches)
     if(NOT _matches EQUAL 1)
       message(FATAL_ERROR
         "Failed to find exported target names in '${_export_file}'")
     endif()
-    string(LENGTH "${_foreach_targets}" _length)
-    math(EXPR _substring_length "${_length} - 25")
-    string(SUBSTRING "${_foreach_targets}" 24 ${_substring_length} _targets)
+    string(REGEX REPLACE "${_regex}" "\\1" _targets "${_foreach_targets}")
     string(REPLACE " " ";" _targets "${_targets}")
     list(LENGTH _targets _length)
 

--- a/ament_cmake_export_interfaces/cmake/ament_cmake_export_interfaces-extras.cmake.in
+++ b/ament_cmake_export_interfaces/cmake/ament_cmake_export_interfaces-extras.cmake.in
@@ -5,7 +5,24 @@ set(_exported_interfaces "@_AMENT_CMAKE_EXPORT_INTERFACES@")
 # include all exported interfaces
 if(NOT _exported_interfaces STREQUAL "")
   foreach(_interface ${_exported_interfaces})
-    include("${@PROJECT_NAME@_DIR}/${_interface}Export.cmake")
-    list(APPEND @PROJECT_NAME@_INTERFACES "@PROJECT_NAME@::${_interface}")
+    set(_export_file "${@PROJECT_NAME@_DIR}/${_interface}Export.cmake")
+    include("${_export_file}")
+
+    # extract the target names associated with the export
+    file(
+      STRINGS "${_export_file}" _foreach_targets
+      REGEX "foreach\\(_expectedTarget (.+)\\)")
+    list(LENGTH _foreach_targets _matches)
+    if(NOT _matches EQUAL 1)
+      message(FATAL_ERROR
+        "Failed to find exported target names in '${_export_file}'")
+    endif()
+    string(LENGTH "${_foreach_targets}" _length)
+    math(EXPR _substring_length "${_length} - 25")
+    string(SUBSTRING "${_foreach_targets}" 24 ${_substring_length} _targets)
+    string(REPLACE " " ";" _targets "${_targets}")
+    list(LENGTH _targets _length)
+
+    list(APPEND @PROJECT_NAME@_INTERFACES ${_targets})
   endforeach()
 endif()

